libavformat/rtcenc.c

```diff
@@ -0,0 +1,1763 @@
+    AVIOContext *pb;
+    RTCContext *rtc = s->priv_data;
+
+    if (!extradata || !extradata_size)
+        return 0;
+
+    /* Not H.264 IOSM format, may be annexb etc. */
+    if (extradata_size < 4 || extradata[0] != 1) {
+        if (!ff_avc_find_startcode(extradata, extradata + extradata_size)) {
+            av_log(s, AV_LOG_ERROR, "Format must be ISOM or annexb\n");
+            return AVERROR_INVALIDDATA;
+        }
+        return 0;
+    }
+
+    /* Parse the SPS/PPS in ISOM format in extradata. */
+    pb = avio_alloc_context(extradata, extradata_size, 0, NULL, NULL, NULL, NULL);
+    if (!pb)
+        return AVERROR(ENOMEM);
+
+    version = avio_r8(pb); /* version */
+    avio_r8(pb); /* avc profile */
+    avio_r8(pb); /* avc profile compat */
+    avio_r8(pb); /* avc level */
+    nal_length_size = avio_r8(pb); /* 6 bits reserved (111111) + 2 bits nal size length - 1 (11) */
+    nb_sps = avio_r8(pb); /* 3 bits reserved (111) + 5 bits number of sps */
+
+    if (version != 1) {
+        av_log(s, AV_LOG_ERROR, "Invalid version=%d\n", version);
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    rtc->avc_nal_length_size = (nal_length_size & 0x03) + 1;
+    if (rtc->avc_nal_length_size == 3) {
+        av_log(s, AV_LOG_ERROR, "Invalid nal length size=%d\n", rtc->avc_nal_length_size);
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    /* Read SPS */
+    nb_sps &= 0x1f;
+    if (nb_sps != 1 || avio_feof(pb)) {
+        av_log(s, AV_LOG_ERROR, "Invalid number of sps=%d, eof=%d\n", nb_sps, avio_feof(pb));
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    rtc->avc_sps_size = avio_rb16(pb); /* sps size */
+    if (rtc->avc_sps_size <= 0 || avio_feof(pb)) {
+        av_log(s, AV_LOG_ERROR, "Invalid sps size=%d, eof=%d\n", rtc->avc_sps_size, avio_feof(pb));
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    rtc->avc_sps = av_malloc(rtc->avc_sps_size);
+    ret = avio_read(pb, rtc->avc_sps, rtc->avc_sps_size); /* sps */
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to read sps, size=%d\n", rtc->avc_sps_size);
+        goto end;
+    }
+
+    /* Read PPS */
+    nb_pps = avio_r8(pb); /* number of pps */
+    if (nb_pps != 1 || avio_feof(pb)) {
+        av_log(s, AV_LOG_ERROR, "Invalid number of pps=%d, eof=%d\n", nb_pps, avio_feof(pb));
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    rtc->avc_pps_size = avio_rb16(pb); /* pps size */
+    if (rtc->avc_pps_size <= 0 || avio_feof(pb)) {
+        av_log(s, AV_LOG_ERROR, "Invalid pps size=%d, eof=%d\n", rtc->avc_pps_size, avio_feof(pb));
+        ret = AVERROR_INVALIDDATA;
+        goto end;
+    }
+
+    rtc->avc_pps = av_malloc(rtc->avc_pps_size);
+    ret = avio_read(pb, rtc->avc_pps, rtc->avc_pps_size); /* pps */
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to read pps, size=%d\n", rtc->avc_pps_size);
+        goto end;
+    }
+
+end:
```

Comment:
Should av_freep(&rtc->avc_sps) here?

Reply:
No, it will not be freed in the rtc_deinit. It is not a function-level variable but rather belongs to the entire muxer.